### PR TITLE
add interface name validation

### PR DESCRIPTION
**Pull request by mars1024**
Interface name in linux is not unrestricted, so bad interface name will break CNI plugin.
This pr will add interface name validation to libcni and skel.

validation rules ref to [https://github.com/torvalds/linux/blob/master/net/core/dev.c#L1024](https://github.com/torvalds/linux/blob/master/net/core/dev.c#L1024)

### CODE REVIEW

File: libcni/api_test.go [hunk 1]
```diff
@@ -1027,13 +1027,48 @@ var _ = Describe("Invoking plugins", func() {
...
+
+					_, err := cniConfig.AddNetworkList(ctx, netConfigList, runtimeConfig)
+					Expect(err).To(Equal(&types.Error{
+						Code:    types.ErrInvalidEnvironmentVariables,
+						Msg:     "interface name is empty",
+						Details: "",
+					}))
+				})
+
+				It("returns the error when interface name is overflow", func() {
```
Comment (jellonek) [inline libcni/api_test.go]: is too long

File: pkg/skel/skel_test.go [hunk 2]
```diff
@@ -119,12 +119,38 @@ var _ = Describe("dispatching to the correct callback", func() {
...
 			Expect(err).To(Equal(&types.Error{
-				Code:    4,
+				Code:    types.ErrInvalidEnvironmentVariables,
 				Msg:     "invalid characters in containerID",
 				Details: "some-%%container-id",
 			}))
 		})
 
+		Context("return error when interface name is invalid", func() {
+			It("returns the error when interface name is overflow", func() {
```
Comment (jellonek) [inline pkg/skel/skel_test.go]: ditto

File: pkg/skel/skel_test.go [hunk 3]
```diff
@@ -119,12 +119,38 @@ var _ = Describe("dispatching to the correct callback", func() {
...
+			})
+
+			It("returns the error when interface name contains invalid characters", func() {
+				environment["CNI_IFNAME"] = "test/test"
+
+				err := dispatch.pluginMain(cmdAdd.Func, cmdCheck.Func, cmdDel.Func, versionInfo, "")
+				Expect(err).To(HaveOccurred())
+				Expect(err).To(Equal(&types.Error{
+					Code:    types.ErrInvalidEnvironmentVariables,
+					Msg:     "interface name contains / or whitespace characters",
```
Comment (jellonek) [inline pkg/skel/skel_test.go]: It would be nice to have two separate tests there for both cases, but I do not insist on that.

Comment (mars1024) [reply pkg/skel/skel_test.go]: updated

File: libcni/api_test.go (hunk 1 above)
Comment (mars1024) [reply libcni/api_test.go]: updated

File: pkg/skel/skel_test.go (hunk 2 above)
Comment (mars1024) [reply pkg/skel/skel_test.go]: updated

**Comment by mars1024** [comment]
@jellonek all updated, ptal, thanks ~

File: pkg/utils/utils.go [hunk 4]
```diff
@@ -49,3 +56,21 @@ func ValidateNetworkName(networkName string) *types.Error {
...
+// 1. The name must not be empty
+// 2. The name must be less than 16 characters
+// 3. The name must not contain / or any whitespace characters
+// ref to https://git.kernel.org/pub/scm/network/iproute2/iproute2.git/tree/lib/utils.c?id=1f420318bda3cc62156e89e1b56d60cc744b48ad#n827
+func ValidateInterfaceName(ifName string) *types.Error {
+	if len(ifName) == 0 {
+		return types.NewError(types.ErrInvalidEnvironmentVariables, "interface name is empty", "")
+	}
+	if len(ifName) > maxInterfaceNameLength {
+		return types.NewError(types.ErrInvalidEnvironmentVariables, "interface name is overflow", fmt.Sprintf("interface name length should be less than %d characters", maxInterfaceNameLength+1))
```
Comment (jellonek) [inline pkg/utils/utils.go]: Please use there: "is too long" instead of "is overflow".

Comment (mars1024) [reply pkg/utils/utils.go]: got it

Comment (mars1024) [reply pkg/utils/utils.go]: done

Comment (mars1024) [reply pkg/utils/utils.go]: @jellonek  updated, ptal, thanks ~

File: pkg/utils/utils.go [hunk 5]
```diff
@@ -49,3 +56,21 @@ func ValidateNetworkName(networkName string) *types.Error {
...
+// 3. The name must not contain / or any whitespace characters
+// ref to https://git.kernel.org/pub/scm/network/iproute2/iproute2.git/tree/lib/utils.c?id=1f420318bda3cc62156e89e1b56d60cc744b48ad#n827
+func ValidateInterfaceName(ifName string) *types.Error {
+	if len(ifName) == 0 {
+		return types.NewError(types.ErrInvalidEnvironmentVariables, "interface name is empty", "")
+	}
+	if len(ifName) > maxInterfaceNameLength {
+		return types.NewError(types.ErrInvalidEnvironmentVariables, "interface name is too long", fmt.Sprintf("interface name should be less than %d characters", maxInterfaceNameLength+1))
+	}
+	if bytes.ContainsAny([]byte(ifName), `/ `) {
```
Comment (jellonek) [inline pkg/utils/utils.go]: Sorry that I didn't caught that earlier, but what about new line, tab or vertical tab (I mean - rest of whitespaces)?
Could use there https://golang.org/pkg/unicode/#White_Space ?

Comment (mars1024) [reply pkg/utils/utils.go]: @jellonek Thanks for your suggestion, updated!

File: pkg/utils/utils.go [hunk 6]
```diff
@@ -49,3 +57,24 @@ func ValidateNetworkName(networkName string) *types.Error {
...
+// 3. The name must not contain / or any whitespace characters
+// ref to https://git.kernel.org/pub/scm/network/iproute2/iproute2.git/tree/lib/utils.c?id=1f420318bda3cc62156e89e1b56d60cc744b48ad#n827
+func ValidateInterfaceName(ifName string) *types.Error {
+	if len(ifName) == 0 {
+		return types.NewError(types.ErrInvalidEnvironmentVariables, "interface name is empty", "")
+	}
+	if len(ifName) > maxInterfaceNameLength {
+		return types.NewError(types.ErrInvalidEnvironmentVariables, "interface name is too long", fmt.Sprintf("interface name should be less than %d characters", maxInterfaceNameLength+1))
+	}
+	for _, r := range bytes.Runes([]byte(ifName)) {
```
Comment (squeed) [inline pkg/utils/utils.go]: I don't think we want to convert to runes here - the iproute2 code is totally not utf-aware. It's basically just doing `uint8` comparisons.

Comment (mars1024) [reply pkg/utils/utils.go]: So you think the original change ```bytes.ContainsAny([]byte(ifName), `/ `)``` is more reasonable? But @jellonek think we should check all whitespace-related characters (like \n, \t ...) I'm a little confused.

Comment (jellonek) [reply pkg/utils/utils.go]: @squeed but what is your suggestion? to list possible characters only from ascii range? or keep original proposal for only two characters

Comment (squeed) [reply pkg/utils/utils.go]: While this is technically slightly different from what `iproute2` does, I think this is probably fine.

**Comment by dcbw** [comment]
For reference, from net/core/dev.c:
```
/**
 *	dev_valid_name - check if name is okay for network device
 *	@name: name string
 *
 *	Network device names need to be valid file names to
 *	to allow sysfs to work.  We also disallow any kind of
 *	whitespace.
 */
bool dev_valid_name(const char *name)
{
	if (*name == '\0')
		return false;
	if (strnlen(name, IFNAMSIZ) == IFNAMSIZ)
		return false;
	if (!strcmp(name, ".") || !strcmp(name, ".."))
		return false;

	while (*name) {
		if (*name == '/' || *name == ':' || isspace(*name))
			return false;
		name++;
	}
	return true;
}
```

**Comment by squeed** [comment]
Aha, interesting. @mars1024 can you block "." and ".." too?

**Comment by jellonek** [comment]
If we want to follow this kernel code - we should also block ":".

**Comment by mars1024** [comment]
@dcbw @squeed @jellonek  Thanks for all your suggestions, I'll update this soon ~

**Comment by mars1024** [comment]
Updated!

**Pull request merged**